README.md
# atlantistalker
Atlantis Talker is a talker based on NUTS. The talker runs as a service on a host machine and users can connect to the service using telnet protocol (e.g. `telnet example.com 7000` when the talker is bound to port 7000 on the host machine).

The code in this repository is a copy of the code published on SourceForge as SVN by one of the original coders nicknamed Spartakus.

The code of the Atlantis Talkers, that runs at `telnet talker.sk 7000`, is not public.


CHANGES
/********************** Made in England, modified in Slovakia ***************
 !! Dolezite zmeny: 1. pridany novy prikaz .pictell (send?) <user> <picture>
                   2. messages & editor - nerobi medzery
                   3. vylepsene .wake - teraz posle aj obrazok
                   4. vypis pri nalogovani - od levelu 2 a vyssie aj site/domain:port
                   5. kill - vypise, kto ta killne + ina hlaska
                   6. promote/demote - detto
                   7. pri nadavkach godom vypise, kto nadaval
                   8. zmenene farby!
		   9. novy prikaz - .godpriv (ak God spusti tak miestnost je
		      PRIVATE aj pre inych godov (snaha o trochu sukromia ;)!!
		      #define GOD_PRIVATE 4
	       10. Board funguje ako ma t.j. zrusil som ten riadok navyse
		      zmena v suboruch write_board, wipe_board, check_messages
	       11. V talker.h som dodefinoval NOEXITS a directory pre
		  POMOCFILES. Pri nabootovani mu vychod "noexits" nebude
		  vadit. takze jail uz frci... *cheers*
	       12. Pridal som prikaz pomoc. Subory s helpami budu v adresari
		  /pomocfiles (zadefinovany v headeri ako POMOCFILES).
		  Mainhelp sa bude nachadzat v subore prvapomoc
   Buko:      13. novy prikaz - .hug - posle obr. hug.pic
              14. novy prikaz - .sos - posle sos wizardom a vyssie
              15. nove prikazy - .suggestion + .complaint + .excuse
              16. + prikaz .notice - cita 15. a zapisuje do suboru noticeboard
              17. + mnohe ine vylepsenia na ktore si nespomeniem... ;))
              18. dlhoocakavany prikaz .kiss - mal by dokonca i fungovat ;)
              19. prikaz .time, nepytajte sa ma co robi
              20. .jail - hodi nespratnika do jailu + zapisuje do suboru ci je
                  dotycny zalarovany alebo nie (JAIL_ROOM=1)
              21. A este super spica prikaz .fight!!! Sance prideluje viac-menej
                  nahodne, ale zavisi to hlavne od levelu. Neviem, ci ma
                  posielat aj nejaky obrazok...
     1.8:     22. fixovanych zopar bugov (hlavne co sa tyka .godpriv ;),
                  pridane farby, vylepseny prikaz .who
              23. C00LTALK - v stadiu pripravy... treba este dorobit funkciu
                  deformator(), ktora veticku troska deformuje... ;)
              24. <Fortune cookies> - vypisuje hlasky zo suboru FORTUNE_FILE
              25. novy prikaz - .think v podstate je to iste co .say, akurat
		  ze robi: "Buko thinks: . o O (This is an example)". Ked je
		  user->muzzled==1 tento prikaz by mohol fungovat... ;))
              26. Pridal som este jeden prikaz: .to
                  Syntax je: .to <user> <text> a skratka je "/", cize ulahcuje
                  zbytocne dlhe vypisovanie mena. Napr.: / pr Ahoj ty vtak!
                  Output je: Buko (to Prizo): Ahoj ty vtak! ;-)
                  NOTE: .to sa mi sekalo s .topic, cize: .to = to .top = topic
     1.9:     27. Fixovanych par bugov (ignore, vypisy, etc...)
              28. Implementovane makra (thanx, Gene ;) => 3 nove prikazy:
                  .macro - prezeranie makier
                  .cmacro - vytvaranie novych makier
                  .dmacro - mazanie makier
		  - s tym suvisia aj zmeny v hlavnej procedure exec_com(),
                  inicializaciu ma na starosti initmacro(); Zapisuju sa do
                  suboru MACRODIR/MACROFILE definovanom v headri.
              29. Odstranil som Copyright (c) by Buko ;-)) hehe
              30. Uprava <ignore> - rovnaky level ignoruje rovnaky a nizzsi
                  level, vyssim levelom sa vypise, ze inak by ignoroval.
maj/jun 97:   31. pridane/odstranene/opat pridane cheaty... ;>
	      32. COUNTER - pocitadlo - pocita pristupy - nalogovania na talker
		  - to aby sme vedeli ze aki sme dobri...
	      33. Dalsia, vyzera to ze definitivna uprava makier - hlavne
		  procedury create() a delete() - teraz pracuju efektivnejsie
     2.0:     34. implementovany cooltalk - vyuziva fciu lamerize() ;-) - rozne
		  upravy s textom (uPrawy TeXtu, FarBY, ()bRateNi3 & iN3...)
	      35. drobne upravy, pridane dalsie fcie pre cooltalk
	      36. teraz 3 nove prikazy: .ctalk .ctopic .cdesc
	      37. Ak niekto pouziva .examine na teba a je <= level ako ty
		  sa ti to vypise.
	      38. Drobne vylepsenia: .shout, record_tell(), exec_command()
	      39. Odhalene a fixovane bugy: cooldesc() - dlzka stringu; hug()
		  a kiss() uz nie je mozny pokial user pouziva editor, nie je
		  mozne menit topic ak user->muzzled.
	      40. funkcne kraviny! - prikaz cow() - vypisuje kraviny
jul/august: moje pokusy o implementaciu Spakkyho prikazov: 
     2.1:     41. sing(user,inpstr) - Xy sings: * lalalalaaa, etc... * (S)
	      42. Backdoor ;) (S)
	      43. dmail() - vylepsene delete mail (S)
	      44. teraz checkuje, ci user pise nadafky v editore (S)
	      45. Kidnap & follow - celkom chutne prikazy ;) (S)
	      46. Pridany prikaz ranks() vypisuje zoznam levelov a pocet prikazov. (S)
	      47. Implementacia meniny(), pridane do long_date(2). (S)
	      48. Prikaz .secho - shoutecho, ako echo ale shoutne. (S)
	      49. JokeBoard stuff: prikazy na pracu s nastenkou na ftipy
		  * write_joke() - napise novy ftip
		  * read_joke() - precita ftipy
		  * delete_joke() - vymaze ftip(y)
		  * pri look() vypise, kolko ftipov je na JokeBoarde (S)
	      50. Udalosti - vypisuje nahodne hlasky/udalosti do miestnosti. (S)
	      51. dalsi prikaz sign() - vypise text do ramceka... (S)
	      52. Bezpecnostna vystraha - posle mail userovi, na ktoreho
		  konto sa niekto pokusal dostat. (S)
	      53. pridany vylepseny design pri who() a people(), uprava v
		  set_desc() (S)
	      54. Novy prikaz: tellall(user,inpstr) - tellne vsetkym dany text. (S)
	      55. zmena v rooms() - zlepsenie, zarovnanie vypisu .rmst sa
		  odteraz vola .rooms (fcia rooms(user,1)) (S)
	      56. Novy sposob podavania requestov!!! Pozri misc_op 9 a 10,
		  posielanie, script posli a account_request(). ;-) (S)

atl-brut.c
#include "atl-head.h"
#include "atl-mydb.h"

void brutalis_wars()
{ 
UR_OBJECT user,u;
RM_OBJECT rm;
int vec,oldway,accel,i,cnt,nah;
/*
userd_memory=get_memory_usage();
sprintf(texthb,"Memory: %d bytes. Delta: %d\n",userd_memory,userd_memory-last_memory);
write_room(testroom,texthb);
last_memory=userd_memory;
*/
 if (tsec%2==1) {
   for(rm=room_first;rm!=NULL;rm=rm->next)
   for(i=0;i<MPVM;i++) if (rm->predmet[i]>-1 && predmet[rm->predmet[i]]->function==24) {
     if (0<=rm->dur[i] && rm->dur[i]<10000) {
       if (tsec%4==0) rm->dur[i]=dec_dur(rm->dur[i],1);
       if (rm->dur[i]==0) {
         if (predmet[rm->predmet[i]]->rdestroy!=NULL) {
           sprintf(text,"%s\n",parse_phrase(predmet[rm->predmet[i]]->rdestroy,NULL,NULL,NULL,0));
           write_room(rm,text);
          }
         rm->predmet[i]=-1;
         rm->dur[i]=0;
         continue;
        }
      }
     else if (rm->dur[i]>9999) {
       for (u=user_first;u!=NULL;u=u->next)
       if (!u->login && u->type!=CLONE_TYPE && u->zaradeny
       && u->socket==(rm->dur[i]%10000)) {
         if (tsec%30==0) {
           rm->dur[i]-=10000;
           if (rm->dur[i]<10000) {
             if (predmet[rm->predmet[i]]->rdestroy!=NULL) {
               sprintf(text,"%s\n",parse_phrase(predmet[rm->predmet[i]]->rdestroy,NULL,NULL,NULL,0));
               write_room(rm,text);
              }
             rm->predmet[i]=-1;
             rm->dur[i]=0;
             break;
            }
          }
  	 if (u->room==rm) break;
         if (u->room->group!=rm->group) { rm->dur[i]=abs_dur(rm->dur[i]); break; }
  	 else if (is_free_in_room(u->room)>-1 && (int)(rand()%3)==0) {
           if (predmet[rm->predmet[i]]->searchphr!=NULL) {
             sprintf(text,"%s\n",parse_phrase(predmet[rm->predmet[i]]->searchphr,u,u,u->room,0));
             write_room(rm,text);
            }
           if (predmet[rm->predmet[i]]->ujoinphr!=NULL) {
             sprintf(text,"%s\n",parse_phrase(predmet[rm->predmet[i]]->ujoinphr,u,u,u->room,0));
             write_user(u,text);
	    }
           if (predmet[rm->predmet[i]]->rjoinphr!=NULL) {
             sprintf(text,"%s\n",parse_phrase(predmet[rm->predmet[i]]->rjoinphr,u,u,u->room,10));
             write_room_except(u->room,text,u);
	    }
	   put_in_room(u->room,rm->predmet[i],rm->dur[i]);
	   rm->predmet[i]=-1;
	   rm->dur[i]=0;
	   break;
          }
        }
      }
     else if (rm->dur[i]<-9999) {
       nah=(-rm->dur[i]%10000);
       u=get_user_by_sock(-rm->dur[i]%10000);
       if (u->room!=rm) {
         rm->dur[i]=abs_dur(rm->dur[i]);
         continue;
        }
       vec=rm->predmet[i];
       nah=0;
       if (predmet[vec]->firerate<0 && tsec%-predmet[vec]->firerate==i%-predmet[vec]->firerate) nah=1;
       else if (predmet[vec]->firerate>0 && rand()%100<predmet[vec]->firerate) nah=1;
       if (nah==1) {
         if (predmet[vec]->ustart!=NULL) {
           sprintf(text,"%s\n",parse_phrase(predmet[vec]->ustart,u,NULL,u->room,0));
           write_user(u,text);
          }
         if (predmet[vec]->rstart!=NULL) {
           sprintf(text,"%s\n",parse_phrase(predmet[vec]->rstart,u,NULL,u->room,10));
           write_room_except(u->room,text,u);
          }
	 rm->dur[i]=dec_dur(rm->dur[i],4+rand()%2);
         hurt(NULL,u,vec);
         if (u->health<10) rm->dur[i]=abs_dur(rm->dur[i]);
         do_alt_funct(u,vec);
	 if (rm->dur[i]==0) {
  	   if (predmet[vec]->udestroy!=NULL) {
	     sprintf(text,"%s\n",parse_phrase(predmet[vec]->udestroy,u,NULL,u->room,0));
	     write_user(u,text);
	    }
	   if (predmet[vec]->rdestroy!=NULL) {
	     sprintf(text,"%s\n",parse_phrase(predmet[vec]->rdestroy,u,NULL,u->room,10));
	     write_room_except(u->room,text,u);
	    }
	   rm->predmet[i]=-1;
	   rm->dur[i]=0;
	  }
        }
      }
    }
  }

for(user=user_first;user!=NULL;user=user->next)
if (!user->login && user->type==USER_TYPE && user->room!=NULL && user->dead==0) {
  if (april) {
    if (april & 16) {
      user->cloak=1+rand()%8;
      if (user->cloak==6) user->cloak=8;
     }
    if (april & 32) user->vis=rand()%2;
    if (april & 64) user->last_login=time(0)-rand()%86400;
    if (april & 256) {
      user->cloak=0;
      user->vis=1;
      user->last_login=time(0);
     }
   }
  if (user->attacking>0) user->attacking--;
  if (tsec%2==1 && user->health<100 && user->heal>0) { user->heal--; user->health++; }
  if (tsec%2==0) {
    if (user->health<100) user->health++;
    if (user->health<10) {
      user->health-=2;
      if (user->health==1) user->health--;
      if (user->health==0) user->dead=14;
     }
   }
  if (user->target>-1) { /* zombie-like entities */
    u=get_user_by_sock(user->target);
    if (u==NULL || (u!=NULL && (u->room!=user->room || u->dead))) {
      user->target=-1;
      continue;
     }
    cnt=0;
    for(i=0;i<MPVM;i++) {
      if (user->room->predmet[i]>-1 && user->room->dur[i]%10000==user->socket
       && predmet[user->room->predmet[i]]->function==24) {
         cnt++;
         vec=user->room->predmet[i];
         accel=0;
	 if (predmet[vec]->firerate<0 && tsec%-predmet[vec]->firerate==i%-predmet[vec]->firerate) accel=1;
	 else if (predmet[vec]->firerate>0 && rand()%100<predmet[vec]->firerate) accel=1;
	 if (accel==1) {
	   if (predmet[vec]->ustart!=NULL) {
	     sprintf(text,"%s\n",parse_phrase(predmet[vec]->ustart,u,NULL,u->room,0));
	     write_user(u,text);
	    }
	   if (predmet[vec]->rstart!=NULL) {
	     sprintf(text,"%s\n",parse_phrase(predmet[vec]->rstart,u,NULL,u->room,10));
	     write_room_except(u->room,text,u);
	    }
	   hurt(user,u,vec);
	   if (u->dead || u->health<10) { user->target=-1; break; }
	   do_alt_funct(u,vec);
	   user->room->dur[i]-=(4+rand()%2)*10000;
	   if (user->room->dur[i]<10000) {
  	     if (predmet[vec]->udestroy!=NULL) {
	       sprintf(text,"%s\n",parse_phrase(predmet[vec]->udestroy,user,u,user->room,0));
	       write_user(user,text);
	      }
	     if (predmet[vec]->rdestroy!=NULL) {
	       sprintf(text,"%s\n",parse_phrase(predmet[vec]->rdestroy,user,u,user->room,10));
	       write_room_except(user->room,text,user);
	      }
	     user->room->predmet[i]=-1;
	     user->room->dur[i]=0;
	    }
	  }
        }
     }
    if (cnt==0) user->target=-1;
   }
  if (user->room->group==4 || (user->into!=NULL && user->into->group==4)) { /* brutalis walking engine */
    if (user->into!=NULL || user->from!=NULL) {
/*      sprintf(text,"%d",user->way);
      write_user(user,text);*/
      vec=get_vehicle(user);
      oldway=user->way;
      if (user->predmet[HANDS+2]>-1) accel=predmet[user->predmet[HANDS+2]]->seconddur;
      else accel=0;
      if (vec>-1) user->way+=10+predmet[vec]->seconddur;
      else user->way+=10+accel;
      if (oldway<0 && user->way>=0) {
        if (user->into==NULL && user->from!=NULL) {
          user->from=NULL;
          if (vec>-1 && predmet[vec]->userphr!=NULL) {
            sprintf(text,"~FG%s\n",parse_phrase(predmet[vec]->userphr,user,NULL,NULL,0));
            write_user(user,text);
           }
          else {
            sprintf(text,"~FGZastal%s si %s.\n",pohl(user,"","a"),user->room->where);
            write_user(user,text);
           }
          if (vec>-1 && predmet[vec]->roomphr!=NULL) {
            sprintf(text,"~FG%s\n",parse_phrase(predmet[vec]->roomphr,user,NULL,NULL,10));
            write_room_except(user->room,text,user);
           }
/*          else {
            sprintf(text,"~FG%s tu zastal%s.\n",pohl(user,"","a"),user->room->where);
            write_room_except(user->room,text,user);
           }*/
         }
        if (user->into!=NULL && user->from!=NULL) {
          user->from=NULL;
         }
       }
      if (user->way>49) {
        if (user->into==NULL) { user->way=0; return; }
        move_vehicle(user,user->into,1);
        user->from=user->room;
        user->room=user->into;
        user->reveal=20+rand()%10;
        user->hidden=0;
        look(user);
        reset_access(user->from);
        if (user->statline==CHARMODE) show_statline(user);
        if (vec>-1 && predmet[vec]->victimphr!=NULL) {
          sprintf(text,"~FG<- %s\n",parse_phrase(predmet[vec]->victimphr,user,NULL,user->from,1));
          write_room_except(user->room,text,user);
         }
        else {
          sprintf(text,"~FG<- %s%s~RS~FW~FG %s.\n",/*teamcolor[user->team]*/"",user->name,user->in_phrase);
          write_room_except(user->room,text,user);
         }
        if (vec>-1) {
          sprintf(text,"~FG-> ~FG%s%s~RS~FG sa stratil%s z dohladu.\n",/*teamcolor[user->team]*/"",user->name,pohl(user,"","a"));
          write_room_except(user->from,text,user);
         }
        else {
          sprintf(text,"~FG-> %s%s~RS~FW~FG %s.\n",/*teamcolor[user->team]*/"",user->name,expand_outphr(user,user->room));
          write_room_except(user->from,text,user);
         }
        user->into=NULL;
        user->way=-40;
        if (user->predmet[HANDS+2]>-1)  {
          user->dur[HANDS+2]--;
          if (user->dur[HANDS+2]==0) {
            if (predmet[user->predmet[HANDS+2]]->udestroy!=NULL) {
              sprintf(text,"%s\n",parse_phrase(predmet[user->predmet[HANDS+2]]->udestroy,user,NULL,NULL,0));
              write_user(user,text);
             }
            if (predmet[user->predmet[HANDS+2]]->rdestroy!=NULL) {
              sprintf(text,"%s\n",parse_phrase(predmet[user->predmet[HANDS+2]]->rdestroy,user,NULL,NULL,10));
              write_room_except(user->room,text,user);
             }
           }
         }
       }
     }
    else user->way=0;
   }
 }
} 

